<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>11_Promise对象</title>
</head>
<body>
<!--
1. 理解:
  * Promise对象: 代表了未来某个将要发生的事件(通常是一个异步操作)
  * 有了promise对象, 可以将异步操作以同步的流程表达出来, 避免了层层嵌套的回调函数(俗称'回调地狱')
  * ES6的Promise是一个构造函数, 用来生成promise实例
2. 使用promise基本步骤(2步):
  * 创建promise对象
    let promise = new Promise((resolve, reject) => {
        //初始化promise状态为 pending
      //执行异步操作
      if(异步操作成功) {
        resolve(value);//修改promise的状态为fullfilled
      } else {
        reject(errMsg);//修改promise的状态为rejected
      }
    })
  * 调用promise的then()
    promise.then(function(
      result => console.log(result),
      errorMsg => alert(errorMsg)
    ))
3. promise对象的3个状态
  * pending: 初始化状态
  * fullfilled: 成功状态
  * rejected: 失败状态
4. 应用:
  * 使用promise实现超时处理

  * 使用promise封装处理ajax请求
    let request = new XMLHttpRequest();
    request.onreadystatechange = function () {
    }
    request.responseType = 'json';
    request.open("GET", url);
    request.send();
-->

<script type="text/javascript">

  setTimeout(() => {
    const sum = (x, y) => x + y;
    sum(1, 2);
    setTimeout(() => {
      sum(3, 4);
      setTimeout(() => {
        sum(5, 6);
      }, 3000)
    }, 2000)
  }, 1000)

  /*
    promise就是一个异步编程的解决方案，用来解决回调地狱问题
   */

  const sum = (x, y) => x + y;

  const promise = new Promise((resolve, reject) => {
    /*
      resolve 解决
      reject 拒绝

      new Promise(), 会创建promise实例对象，实例对象内部默认是pending状态（初始化状态）
      resolve()  将promise状态由初始化状态改为 fullfilled 成功的状态
      reject() 将promise状态由初始化状态改为 rejected 失败的状态
     */
    // 当前函数中一般只放异步代码
    setTimeout(() => {
      // resolve(1); // promise状态只能有初始化状态改为成功/失败的状态。不能由成功变成失败或者失败变成成功
      reject(2); // 将promise状态由初始化状态改为 fullfilled / resolved 成功的状态
    }, 1000)
  });

  // promise实例对象， then方法
  /*promise.then((result) => {
    // 当promise对象状态变成成功的状态时，会调用当前函数
    // 成功的回调函数可以接受resolve方法传入参数
    console.log('成功的回调函数触发了~');
    console.log(result);
    sum(1, 2);
  }, (error) => {
    // 当promise对象状态变成失败的状态时，会调用当前函数
    // 失败的回调函数可以接受reject方法传入参数
    console.log('失败的回调函数触发了~');
    console.log(error);
  })*/

  promise
    .then((result) => {
      // 当promise对象状态变成成功的状态时，会调用当前函数
      // 成功的回调函数可以接受resolve方法传入参数
      console.log('成功的回调函数触发了~');
      console.log(result);
      console.log(promise);
      sum(1, 2);
    })
    .catch((error) => {
      // 当promise对象状态变成失败的状态时，会调用当前函数
      // 失败的回调函数可以接受reject方法传入参数
      console.log('失败的回调函数触发了~');
      console.log(error);
      console.log(promise);
    })

  console.log(promise);




</script>

</body>

</html>